File: C++/124.二叉树中的最大路径和.cpp

// 124. 二叉树中的最大路径和
//
// 路径 被定义为一条从树中任意节点出发，沿父节点-子节点连接，达到任意节点的序列。
// 同一个节点在一条路径序列中 至多出现一次 。该路径 至少包含一个 节点，且不一定经过根节点。
// 路径和 是路径中各节点值的总和。
// 给你一个二叉树的根节点 root ，返回其 最大路径和 。

#include <stdint.h>

#include <algorithm>

#include "header.h"

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int max_sum = INT32_MIN;
        path_sum(root, &max_sum);
        return max_sum;
    }

    int path_sum(TreeNode* root, int* max_sum) {
        if (root == nullptr) {
            return 0;
        }

        // 获取左右子树能够提供的最大值 -- 只有在最大贡献值大于0时, 才会选取对应子节点
        int left_sum = std::max(path_sum(root->left, max_sum), 0);
        int right_sum = std::max(path_sum(root->right, max_sum), 0);

        // 更新最大路径和
        *max_sum = std::max(*max_sum, root->val + left_sum + right_sum);

        // 返回当前节点能够为父节点提供的最大贡献值
        return root->val + std::max(left_sum, right_sum);
    }
};

File: leetcode/06树/124. 二叉树中的最大路径和.go
package main

import "math"

// 124. 二叉树中的最大路径和

// 二叉树中的 路径 被定义为一条节点序列，序列中每对相邻节点之间都存在一条边。同一个节点在一条路径序列中 至多出现一次 。该路径 至少包含一个 节点，且不一定经过根节点。
//
// 路径和 是路径中各节点值的总和。
//
// 给你一个二叉树的根节点 root ，返回其 最大路径和 。

// maxPathSum .
func maxPathSum(root *TreeNode) int {
	maxSum := math.MinInt32
	pathSum(root, &maxSum)
	return maxSum
}

// pathSum .
// 返回给定根节点能够为其父节点提供的最大贡献值（若贡献值为负则返回0）
// 在计算的过程中更新全局最大路径和
func pathSum(root *TreeNode, maxSum *int) int {
	if root == nil {
		return 0
	}

	// 左右子树能为该节点提供的最大贡献 -- 最小为0
	leftSum := pathSum(root.Left, maxSum)
	rightSum := pathSum(root.Right, maxSum)

	// 计算当前节点作为根节点的最大路径和
	priceNewPath := root.Val + leftSum + rightSum

	// 更新全局最大路径和
	*maxSum = max(*maxSum, priceNewPath)

	// 返回当前节点对其父节点的最大贡献值（若贡献值为负则返回0）
	return max(root.Val+max(leftSum, rightSum), 0)
}
